### ISSUE TITLE
Long path bugs (Could not open path)

### ISSUE THREAD

**Issue by AlttiRi** [easymodo/qimgv#290]
## TL;DR
The program works with bugs when I try to open files with the long path.
**In comparison, Windows Photo Viewer opens these file fine.**

## How to test it
Download the attachment to test the program with this files.
The attachment: [unpack me here.zip](https://github.com/easymodo/qimgv/files/6011162/unpack.me.here.zip)

I use the follow file structure:
```
Folders:
[example-folder-0] images with long path 2020.02.20/                                      aka "[0]"
[subfolder-1] some discriptions about the folder/                                         aka "[1]"
[subfolder-2] this folder has the more long name thank to with subfolder/                 aka "[2]"
[subfolder-3] Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod/    aka "[3]"

Files:
[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line multiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png
[example.com] 90897002—admin—2020.01.01—the example image 2—picture white_background simple_background red_line multiple_lines red_lines lorem_ipsum dolor_sit_amet consectetur—7b301e17fc14e42f75d779ca8ecb2dc0.png
[example.com] 90897003—admin—2020.01.01—the example image 3—picture white_background simple_background green_line multiple_lines green_lines lorem_ipsum dolor_sit_amet consectetur—050bc0a8f7637b0a0d0faa781388e3f7.png

aka "1.png",  "2.png",  "3.png" correspondingly. 
```
```
[0]/1.png
[0]/2.png
[0]/3.png
[0]/[1]/1.png
[0]/[1]/2.png
[0]/[1]/3.png
[0]/[2]/[3]/1.png
[0]/[2]/[3]/2.png
[0]/[2]/[3]/3.png
```

## Bugs

When I open files in
- `C:\Users\Username\Downloads\[0]\`
- `C:\[0]\`
- `C:\[0]\[1]`
- `C:\[0]\[2]\[3]`

The program creates one additional virtual file with name `_EXAMP~1.PNG`, but the other files open fine _(I can you the scroll)_

![1](https://user-images.githubusercontent.com/16310547/108528831-e4b35500-72e4-11eb-89fb-8d871811ff56.png)

---

When I open files in 
- `C:\Users\Username\Downloads\[0]\[1]`
- `C:\Users\Username\Downloads\[0]\[2]\[3]`
The program open only one file with name `_EXAMP~1.PNG`. I can scroll to other files, only [ 1/1 ].

![2](https://user-images.githubusercontent.com/16310547/108528842-e715af00-72e4-11eb-883d-3bb07346ff21.png)

---

When I try to open a file on other hard drive:
- `M:\[0]`

I got the error: `Could not open path: \\?\M:\[0]\1.png`
![3](https://user-images.githubusercontent.com/16310547/108529199-4bd10980-72e5-11eb-8a9d-53e42620cadb.png)

**Comment by easymodo** [comment]
yep...

Tried disabling 8dot3 name creation in registry, which of course had no effect

>Windows Photo Viewer opens these file fine.

Probably uses some winapi incantations to convert from 8.3 short names. I may have to to the same but no guarantees

**Comment by easymodo** [comment]
just use shorter names for now or use this if you *really* need to exceed 260 characters https://support.code42.com/CrashPlan/6/Troubleshooting/Windows_file_paths_longer_than_255_characters

**Comment by AlttiRi** [comment]
> https://support.code42.com/CrashPlan/6/Troubleshooting/Windows_file_paths_longer_than_255_characters

It's thing for very special cases.

By the way, node.js works in all cases fine from the box.

```js
#!/usr/bin/env node

const fs = require("fs").promises;
const path = require("path");

!async function main() {
    for await (const file of listFiles()) {
        console.log(file);
        /** @type {Buffer} */
        const jsonText = await fs.readFile(file);
        console.log(jsonText.length);
    }
}();

async function * listFiles(location = process.cwd(), recursively = true) {
    const files = await fs.readdir(location);
    for (const file of files) {
        const filepath = path.resolve(location, file);
        /** @type {Stats} */
        const stat = await fs.stat(filepath);
        if (!stat.isDirectory() || !recursively) {
            yield filepath;
        } else {
            yield * listFiles(filepath);
        }
    }
}

```

<details>
  <summary>Log</summary>

  ```
M:\[example-folder-0] images with long path 2020.02.20\[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line multiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png
14024
M:\[example-folder-0] images with long path 2020.02.20\[example.com] 90897002—admin—2020.01.01—the example image 2—picture white_background simple_background red_line multiple_lines red_lines lorem_ipsum dolor_sit_amet consectetur—7b301e17fc14e42f75d779ca8ecb2dc0.png
13635
M:\[example-folder-0] images with long path 2020.02.20\[example.com] 90897003—admin—2020.01.01—the example image 3—picture white_background simple_background green_line multiple_lines green_lines lorem_ipsum dolor_sit_amet consectetur—050bc0a8f7637b0a0d0faa781388e3f7.png
13691
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-1] some discriptions about the folder\[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line multiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png
14024
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-1] some discriptions about the folder\[example.com] 90897002—admin—2020.01.01—the example image 2—picture white_background simple_background red_line multiple_lines red_lines lorem_ipsum dolor_sit_amet consectetur—7b301e17fc14e42f75d779ca8ecb2dc0.png
13635
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-1] some discriptions about the folder\[example.com] 90897003—admin—2020.01.01—the example image 3—picture white_background simple_background green_line multiple_lines green_lines lorem_ipsum dolor_sit_amet consectetur—050bc0a8f7637b0a0d0faa781388e3f7.png
13691
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-2] this folder has the more long name thank to with subfolder\[subfolder-3] Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod\[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line multiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png
14024
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-2] this folder has the more long name thank to with subfolder\[subfolder-3] Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod\[example.com] 90897002—admin—2020.01.01—the example image 2—picture white_background simple_background red_line multiple_lines red_lines lorem_ipsum dolor_sit_amet consectetur—7b301e17fc14e42f75d779ca8ecb2dc0.png
13635
M:\[example-folder-0] images with long path 2020.02.20\[subfolder-2] this folder has the more long name thank to with subfolder\[subfolder-3] Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod\[example.com] 90897003—admin—2020.01.01—the example image 3—picture white_background simple_background green_line multiple_lines green_lines lorem_ipsum dolor_sit_amet consectetur—050bc0a8f7637b0a0d0faa781388e3f7.png
13691

```

  
</details>

**Comment by AlttiRi** [comment]
_Just for clarification, I have enabled "NTFS long paths policy"._
https://superuser.com/a/1119980/1034786

**Comment by AlttiRi** [comment]
Wait, but Qt works fine with long paths, I have checked it right now. 
Did I miss something?

```c++
#include <QMessageBox>
#include <QFileInfo>
#include <QFile>

// ...

QString pathStr = "C:/[example-folder-0] images with long path 2020.02.20/[subfolder-2] this folder has the more long name thank to with subfolder/[subfolder-3] Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod/";
QString fileStr = "[example.com] 90897003—admin—2020.01.01—the example image 3—picture white_background simple_background green_line multiple_lines green_lines lorem_ipsum dolor_sit_amet consectetur—050bc0a8f7637b0a0d0faa781388e3f7";


QFile file(pathStr + fileStr + ".txt");
if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
    return;

QByteArray line = file.readLine();
QMessageBox::information(this, "Text file line", line); // ok. it shows the first line of the txt file

QFileInfo fileInfo(pathStr + fileStr + ".png");
if (fileInfo.exists() && fileInfo.isFile()) {
    QMessageBox::information(this, "File exists, size:", QString::number(fileInfo.size()));  // ok. it shows the size of the png file
} else {
    QMessageBox::information(this, "File does not exists", "");
}

```

**Comment by easymodo** [comment]
the problem is that we already get the useless shortened path via commandline argument (when opening from explorer), and also from drag'n'drop

i think the way to go is guessing via regex if the path is short version, then convert it via winapi

**Comment by AlttiRi** [comment]
> When I try to open a file in other hard drive:
> 
> * `M:\[0]`
> 
> I got the error: `Could not open path: \\?\M:\[0]\1.png` ("\\\\?\\" is [verbatim](https://docs.microsoft.com/ru-ru/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#win32-file-namespaces)).
> ![3](https://user-images.githubusercontent.com/16310547/108529199-4bd10980-72e5-11eb-8a9d-53e42620cadb.png)

In this case the program gets the correct, not shorted, path.

![Screenshot](https://user-images.githubusercontent.com/16310547/110222688-21866b00-7ee5-11eb-982f-b0f5005039fc.png)

It just only has the additional verbatim at the string start (`\\?\`).

_Damn, I have already deinstalled Qt, can't check this case._

**Comment by AlttiRi** [comment]
Hm, there is an interesting moment.

I have bash script that prints input file path (I just drag'n'drop it on the scripts):
```bash
#!/bin/bash

# Input file
echo "$1"

echo ""; read -p "Press enter to continue..."
```

When I drop on it the file from disk `C` I get:
`C:\_EXAMP~1.20\_SUBFO~1\_EXAMP~1.PNG`

But **I can't drop** the file with long path from other disk:
![Screenshot](https://user-images.githubusercontent.com/16310547/110224367-04a16600-7eec-11eb-80e3-9279fb9cb999.png)


I have no assumption why.

**Comment by AlttiRi** [comment]
I have reinstalled Qt to recheck that long path.

Yes, it works too fine.
Both
```c++
QString pathStr = "M:/[example-folder-0]...
```
and
```c++
QString pathStr = "\\\\?\\M:\\[example-folder-0] ...
```

It strange that the program shows in the exception pop up **the correct path**, but does not work at all.

With shorted paths it works, but with bugs, as I described in the first message.

**Comment by AlttiRi** [comment]
**Wait, maybe the problem reason is the old Qt?** You use Qt **5.12.14** (Prog version is 0.9.1)

Here is the program's code:
https://github.com/easymodo/qimgv/blob/a1058effeb341dfccab549a53c1e0be7c2c03415/qimgv/core.cpp#L813-L828

It works fine with the latest **Qt 6.0.2**:
```c++
QString pathStr = "\\\\?\\M:\\[example-folder-0] images with long path 2020.02.20\\";
QString fileStr = "[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line multiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png";
QString path = pathStr + fileStr;

QFileInfo fileInfo(path);
QString directoryPath;
if(fileInfo.isDir()) {
    directoryPath = QDir(path).absolutePath();
} else if(fileInfo.isFile()) {
    directoryPath = fileInfo.absolutePath();
} else {
    qDebug() << "Could not open path: " << path;
    QMessageBox::information(this, "Could not open path", path);
    return;
}

qDebug() << "directoryPath: " << directoryPath;
qDebug() << "fileInfo.size: " << fileInfo.size();
QMessageBox::information(this, "Open path", path);
```

---

My output:

![Screenshot](https://user-images.githubusercontent.com/16310547/110345720-f06e7d80-803f-11eb-881e-0e32bdde3e68.png)

The program result:

![110222688-21866b00-7ee5-11eb-982f-b0f5005039fc](https://user-images.githubusercontent.com/16310547/110345816-02e8b700-8040-11eb-98ce-98356019f2eb.png)

**The path strings are the same.**

I think it should works with Qt 6.0.2.

**Comment by easymodo** [comment]
Qt version probably has nothing to do with this. I'm using `std::filesystem` for directory listing (performance reasons), which is by the way giving me some weird issues..

For example when i iterate through the zip file you provided it fails to read some files, while reading others just fine (even though they are sitting right next to each other in a folder). Still looking into it

**Comment by AlttiRi** [comment]
Yes, it's Qt version's problem.

I have installed [Qt 5.2.1](https://www.qt.io/blog/2014/02/05/qt-5-2-1-released).
Now I get the error when I'm trying to open a long path.

![Screenshot](https://user-images.githubusercontent.com/16310547/110375539-2709bf80-8063-11eb-84a7-196f844cf580.png)


There was no problem with Qt 6.0.2.

**Comment by AlttiRi** [comment]
Please, note, there are **two types** of problems:
- with **long paths** `"\\\\?\\M:\\[example-folder-0] images with long path 2020.02.20\\";`
- with **short paths** `"C:\\Users\\Username\\Downloads\\_EXAMP~1.20\\_SUBFO~2\\_SUBFO~1"`

Files with long paths currently do no open at all. This problem can be fixed with the newer Qt version.
Files with "short" paths opens, but with bugs. (The problem examples is in the first message.)

**Comment by AlttiRi** [comment]
>I'm using std::filesystem for directory listing 

By the way, Qt 6 uses C++17.
In C++17 there are some changes about `std::filesystem`, and _probably_ some changes are about working with long paths in Windows.

**Comment by AlttiRi** [comment]
By the way, you have added Qt6 https://github.com/easymodo/qimgv/commit/d0fe882cf5b763a22962106ae2b00d416dd8d2e7
I can't check it, since it's hard to build it for me, and the last alpha build was posted before this commit.

Did it affect at that bugs in any way?

**Comment by easymodo** [comment]
I haven't tried qt6 on windows yet

**Comment by AlttiRi** [comment]
Well, with 1.0.3 no any problem is resolved.

But not it does not display the floated warning `Could not open path: \\?\D:\...`

**Comment by AlttiRi** [comment]
Eh, wait. It's still Qt 5.15.2.

I thought you have updated it to Qt 6 some months ago.

**Comment by AlttiRi** [comment]
I still believe that Qt 6 can help.

**Comment by AlttiRi** [comment]
Another example that Qt works fine with long paths:

```c++
#include "mainwindow.h"

#include <QApplication>
#include <QMessageBox>
#include <QFileInfo>

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);

    qDebug() << "argc: " << argc;
    qDebug() << "argv[0]: " << argv[0];
    qDebug() << "argv[1]: " << argv[1];

    QMessageBox msgBox;
    msgBox.setWindowTitle("argv[0]");
    msgBox.setText(argv[0]);
    msgBox.exec();

    if (argc == 2) {
        QMessageBox msgBox2;
        msgBox2.setWindowTitle("argv[1]");
        msgBox2.setText(argv[1]);
        msgBox2.exec();

        QFileInfo fileInfo(argv[1]);
        qDebug() << "fileInfo.size: " << fileInfo.size();

        QMessageBox msgBox3;
        msgBox3.setWindowTitle("fileInfo.size");
        QString size = QString::number(fileInfo.size());
        msgBox3.setText(size);
        msgBox3.exec();
    }

    return 0;
}
```

Build the exe file. Set up in File Explorer opening of files (for example, `.png`) with this exe by a double click. Then open a file with a long path with this exe by a double click.

It works.

On disk C:
![image](https://user-images.githubusercontent.com/16310547/148742423-d033ddaf-8cb8-4b18-a51d-f7b1fa972712.png)

On non-C disk:
![image](https://user-images.githubusercontent.com/16310547/148742533-d1507824-9bba-4aed-87f9-1c9263294260.png)

In both cases the file (size of it) was successfully read:
![image](https://user-images.githubusercontent.com/16310547/148742769-2a6e19b0-0360-45b9-89c1-5255010d300a.png)

But qimgv currently does not work with files with a long path on non-C disk.

**Comment by AlttiRi** [comment]
BTW, in Node.js there is [`fs.realpath`](https://nodejs.org/api/fs.html#fspromisesrealpathpath-options). Possibly there is such function in Qt/C++.

```js
const dir = "C:\\#LONGP~1\\_DIREC~1\\LOREMI~1.UTE";
const filepath = dir + "\\" + "_EXAMP~1.PNG";

console.log(filepath); 
// C:\#LONGP~1\_DIREC~1\LOREMI~1.UTE\_EXAMP~1.PNG

console.log(await fs.realpath(filepath));
// C:\#longpath - directory with a file with long path\_directory with a file with long path\Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam\[example.com] 90897001—admin—2020.01.01—the example image 1—picture white_background simple_background black_line ultiple_lines black_lines lorem_ipsum dolor_sit_amet consectetur—5553a3829fbae129937638c5f0acc8cd.png

```

**Comment by AlttiRi** [comment]
I have tested it again.

Finally. I think it can be fixed easily.

How? 
Use https://github.com/gulrak/filesystem instead of the default `<filesystem>`.

---

Here is the my test repo: https://gist.github.com/AlttiRi/f2781c1d60ec482f3c50c26703072b45
(Only add `filesystem.hpp` from https://github.com/gulrak/filesystem/releases/tag/v1.5.10)

---

Qt works fine with long paths, no need even to add `\\?\` (`\\?\UNC\`):
```c++
QPixmap pic(imgPathLong2);
ui->label_5->setPixmap(pic);
```

But since you use `std::filesystem` it is most likely the problem reason.
```c++
qDebug() << "std::filesystem::exists:" << std::filesystem::exists(imgPathLong2.toStdString());
```
It prints `false`.

But there is https://github.com/gulrak/filesystem.

Just include it and replace `std` with `ghc`:
```c++
qDebug() << "ghc::filesystem::exists:" << ghc::filesystem::exists(imgPathLong2.toStdString());
```
It prints `true`.

![image](https://user-images.githubusercontent.com/16310547/149612764-155d82e4-b959-4ed0-8f57-08071eb18caf.png)
![image](https://user-images.githubusercontent.com/16310547/149612769-21b7be15-f8ec-4671-8f53-d744701bfcc3.png)

**Comment by AlttiRi** [comment]
To convert 8.3 names to long filenames just use [`GetLongPathNameA`](https://docs.microsoft.com/windows/win32/api/fileapi/nf-fileapi-getlongpathnamea).

Here is my code (verify it):
```c++

QString longFilePathA(QString filepath, DWORD cchBuffer = 260)
{
    LPCSTR lpszShortPath = filepath.toStdString().c_str();
    LPSTR lpszLongPath = new char[cchBuffer];

    DWORD length = GetLongPathNameA(lpszShortPath, lpszLongPath, cchBuffer);
    qDebug() << "length:" << length;

    if (length > cchBuffer) {
        return longFilePathA(filepath, length);
    }
    return QString(lpszLongPath);
}
```



The result:
![image](https://user-images.githubusercontent.com/16310547/149617706-c0f2fdd3-3e23-4a42-9a83-d0c65d59907f.png)

**Comment by AlttiRi** [comment]
So, both problems are solvable.

1. Use [`ghc::filesystem`](https://github.com/gulrak/filesystem) with support of long paths. 
(Or why not just use `QDir` and etc? _Is it really worth to use `filesystem`'s functions instead of `Qt`'s ones?_)
2. Use [`GetLongPathNameA`](https://docs.microsoft.com/windows/win32/api/fileapi/nf-fileapi-getlongpathnamea) to convert 8.3 name to long name.

**Comment by easymodo** [comment]
>why not just use QDir

Bad performance. Set sorting to "date" on large directories and you'll see. But that was long time ago, may have been fixed in qt, need to test

Anyway, i've just tried this ghc::filesystem - it works except in some paths i need to do 'GetLongPathNameA"' as you said

![image](https://user-images.githubusercontent.com/5484275/149618273-0bfb5864-84d5-403a-ae48-07648af27591.png)

**Comment by AlttiRi** [comment]
It looks `GetLongPathNameA` does not works with Non-ASCII chars.

So or use it _only_ with short names, or use `GetLongPathNameW` (did not test it) for any file path.

**Comment by AlttiRi** [comment]
> Set sorting to "date" on large directories and you'll see. 

BTW, the sorting by date currently works with bugs if the images are created within a short time.
https://github.com/easymodo/qimgv/issues/362

**Comment by AlttiRi** [comment]
Here is it, `GetLongPathNameW` (with Unicode support):
```c++
QString longFilePathW(QString filepath, DWORD cchBuffer = 512)
{
    LPCWSTR lpszShortPath = (LPCWSTR) filepath.utf16();
    LPWSTR lpszLongPath = new WCHAR[cchBuffer];

    DWORD length = GetLongPathNameW(lpszShortPath, lpszLongPath, cchBuffer);
    qDebug() << "length:" << length;

    if (length > cchBuffer) {
        delete [] lpszLongPath;
        return longFilePathW(filepath, length);
    }

    QString result = QString::fromStdWString(lpszLongPath);
    delete [] lpszLongPath;
    return result;
}
```

**Comment by AlttiRi** [comment]
Yes, sorting by the date takes time, since it requires to call [`stat`](https://en.wikipedia.org/wiki/Stat_(system_call)) for each file to get the file info.

But I see no speed difference between the approach with Qt and how qimgv currently works.
For example, sorting of 11k of images takes 1 second with Qt, qimgv opens a file from this directory also for ~1+ second.

See the related bug: https://github.com/easymodo/qimgv/issues/383

Here is my code:
```c++

    QString dirName = "C:\\FolderWithImages\\";
    qDebug() << dirName;

    QDir testDir = QDir(dirName);
    testDir.setFilter(QDir::Files | QDir::NoDotAndDotDot);

    QList<QFileInfo> fileInfoList = testDir.entryInfoList();
    qDebug() << fileInfoList.size();


    QElapsedTimer timer1;
    timer1.start();
    qDebug() << "by lastModified" << fileInfoList.at(0);
    std::sort(fileInfoList.begin(), fileInfoList.end(),
              [](const QFileInfo& a, const QFileInfo& b) {
                  return a.lastModified() < b.lastModified(); // .birthTime()
              }
    );
    qDebug() << "by lastModified" << fileInfoList.at(0);
    qDebug() << "timer1:" << timer1.elapsed() << "ms";

```

**Comment by AlttiRi** [comment]
BTW, there is [`QFileInfo::setCaching(bool enable)`](https://doc.qt.io/qt-6/qfileinfo.html#setCaching), [`QFileInfo::stat()`](https://doc.qt.io/qt-6/qfileinfo.html#stat), but I don't understand why this code does not work:

```c++
    for (auto const& fileInfo : fileInfoList) {
        fileInfo.setCaching(true);
        // or
        fileInfo.stat();
    }
```

The functions can speed up the code as I assume.

But I get:
```
error: 'this' argument to member function 'setCaching' has type 'const QFileInfo', but function is not marked const
```

**Comment by AlttiRi** [comment]
Since sorting by the date takes 2 ms and
```c++
    QElapsedTimer timerGoTakeMtime;
    timerGoTakeMtime.start();
    for (auto &fileInfo : fileInfoList) {
        fileInfo.lastModified();
    }
    qDebug() << "timerGoTakeMtime:" << timerGoTakeMtime.elapsed() << "ms";
```
takes ~30 ms.

It makes sense to create a class with `filepath` property and some fields from `stat` system call.

It looks that `QFileInfo` calls `stats` every time in the `sort`'s callback, (As well as your current code (I assume), since it has the same speed.)

**Comment by AlttiRi** [comment]
**50 ms instead of 1000 ms** to sort by mtime:

```C++
    QElapsedTimer timerX;
    timerX.start();

    class FileEntry {
    public:
        QString filenath;
        QDateTime mtime;
        FileEntry(QString filenath, QDateTime mtime) {
            this->filenath = filenath;
            this->mtime = mtime;
        }
    };

    QList<FileEntry> fileEntryList;
    for (QFileInfo &fileInfo : fileInfoList) {
        QString filenath = fileInfo.absoluteFilePath();
        QDateTime mtime = fileInfo.lastModified();
        fileEntryList << FileEntry(filenath, mtime);
    }

    std::sort(fileEntryList.begin(), fileEntryList.end(),
              [](const FileEntry& a, const FileEntry& b) {
                  return a.mtime < b.mtime;
              }
    );
    qDebug() << "fileEntryList.size():" << fileEntryList.size();
    qDebug() << "fileEntryList.at(0).name:" << fileEntryList.at(0).filenath;
    qDebug() << "timerX:" << timerX.elapsed() << "ms";
```

**Comment by AlttiRi** [comment]
Okay, I did it.

The full working demo: https://github.com/AlttiRi/demo-image-viewer

Take a look at the readme.

---

It has fixed this bug: just call `GetLongPathNameW` on user input paths:  `argv[1]` and the path from drag'n'drop.

Also it performs the parsing of a directory **multiple times faster** that you did. 

**Qimgv takes 2800 ms** to show an image in a folder with 30k of files, while it's possible to show the image for **150 ms** and **250 ms** to parse the directory' files for sorting by size, mtime and **btime**.

Qimgv currently does not support sorting by `btime` because of std::filesystem.

**Comment by AlttiRi** [comment]
Also Qimgv does not update the title when I toggle/change the sorting.

I mean: if an image is `2/1000`, after the toggling the sorting the window title should contain `999/1000` text.

**Comment by AlttiRi** [comment]
Also it will fix this issue https://github.com/easymodo/qimgv/issues/396

I have checked my demo program, it works fine with UNC paths.

**Comment by AlttiRi** [comment]
I still would like if you add some features that I have implemented in this demo image viewer:
https://github.com/AlttiRi/demo-image-viewer

![Screenshot](https://user-images.githubusercontent.com/16310547/191068071-ef25fbb8-0cc2-4a07-93eb-fab3e1b2c5fe.png)
(1st and 4th items are related each other)

**Comment by AKL55** [comment]
Would appreciate long paths support for windows as i stopped using qimgv because of this issue.
Other alternatives are not as fast.
